### DISCUSSION TITLE
How to generate persistent layout with getStaticProps at build time

### DISCUSSION THREAD

**Discussion by dontfireme** [vercel/next.js#14501] [Help]
I am using this pattern for building persistent layouts: (https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs) [Persistent Layout Patterns in Next.js][1]

It works fine when your layout doesn't rely on external data.

What to do if you want to build a documentation like page - and in your layout there is a sidebar with sections like `Getting Started`, `Examples`, `FAQ`, `About`? Ideally I would like to store those sections in a markdown file and generate the layout at the build time.

    // markdown file
    ---
    section_1: "Getting Started"
    section_2: "Examples"
    section_3: "FAQ"
    section_4: "About"
    ---

How to achieve such functionality with Next.JS and `getStaticProps`?

**Comment by jamesmosier** [comment]
The [blog-starter example](https://github.com/vercel/next.js/blob/canary/examples/blog-starter/lib/api.js) has some good ideas for this.

Basically you'd need to read the markdown file within `getStaticProps`:

```js
import fs from 'fs'
import { join } from 'path'

export async function getStaticProps() {
  const fileDirectory = join(process.cwd(), 'data') // assuming your file is in `data/nav.md`
  const filePath = join(fileDirectory, 'nav.md')
  const fileContents = fs.readFileSync(filePath, 'utf8')
  // now you can parse `fileContents` which will be the contents of your markdown file
  const data = <the data you parse from the file>
  return { props: { data } }
}
```

In the linked example it uses the [grey-matter](https://www.npmjs.com/package/gray-matter) npm package to parse the markdown front-matter so you could do something similar.

**Comment by tettoffensive** [comment]
I'm having a similar issue where my Layout component keeps refreshing on route navigation (whole screen goes blank for a few seconds and comes back).

```ts
export default class App extends NextApp {
  render() {
    const { Component, pageProps, router } = this.props;
    return (
      <FirebaseAuthProvider>
        <SWRConfig
          value={{
            fetcher: fetch,
          }}
        >
          <CacheProvider value={cache}>
            {globalStyles}
            <Layout>
              <Component {...pageProps} key={router.route} />
            </Layout>
          </CacheProvider>
        </SWRConfig>
      </FirebaseAuthProvider>
    );
  }
}
```

My site is not a static blog, but functions more like an app and my pages use `getServerSideProps`.

**How can I get the layout which has a navbar with the username to not refresh on every navigation?**

**Reply by tettoffensive** [reply]
It seems to me that this is only happening on dynamic routes.

I have `/pages/[user]/index.ts` and `/pages/[user]/[...blocks].ts` When I use the same code for my pages for `/pages/test1.ts` and `pages/test2.ts` the whole page refresh doesn't happen.

**Reply by jamesmosier** [reply]
You likely are not using [next/link](https://nextjs.org/docs/api-reference/next/link#dynamic-routes) for dynamic routes correctly. It is hard to say without seeing more code though.

**Reply by songkeys** [reply]
> I have /pages/[user]/index.ts and /pages/[user]/[...blocks].ts When I use the same code for my pages for /pages/test1.ts and pages/test2.ts the whole page refresh doesn't happen.

@tettoffensive I have exactly the same problem! May i ask did you fix it?

The page doesn't refresh if i remove the `getServerSideProps` function (even though it's just an empty function!). Sorry the code base is too large to share with. But I'll try provide a minimal production later
